doc/source/drivers/raster/gti.rst

```diff
@@ -0,0 +1,427 @@
+.. _raster.gti:
+
+================================================================================
+GTI -- GDAL Raster Tile Index
+================================================================================
+
+.. versionadded:: 3.9
+
+.. shortname:: GTI
+
+.. built_in_by_default::
+
+Introduction
+------------
+
+The GTI driver is a driver that allows to handle catalogs with a large
+number of raster files (called "tiles" in the rest of this document, even if a
+regular tiling is not required by the driver), and build a virtual mosaic from
+them. Each tile may be in any GDAL supported raster format, and be a file
+stored on a regular filesystem, or any GDAL supported virtual filesystem (for
+raster drivers that support such files).
+
+This driver offers similar functionality as the :ref:`VRT <raster.vrt>`
+driver with the following main differences:
+
+* The tiles are listed as features of any GDAL supported vector format. Use of
+  formats with efficient spatial filtering is recommended, such as
+  :ref:`GeoPackage <vector.gpkg>`, :ref:`FlatGeoBuf <vector.flatgeobuf>` or
+  :ref:`PostGIS <vector.pg>`. The GTI driver can thus use a larger number of
+  tiles than the VRT driver (hundreds of thousands or more), provided the
+  underlying vector format is efficient.
+
+* The tiles may have different SRS. The GTI driver is capable of on-the-fly
+  reprojection
+
+* The GTI driver offers control on the order in which tiles are composited,
+  when they overlap (z-order)
+
+* The GTI driver honours the mask/alpha band when compositing together
+  overlapping tiles.
+
+* Contrary to the VRT driver, the GTI driver does not enable to alter
+  characteristics of referenced tiles, such as their georeferencing, nodata value,
+  etc. If such behavior is desired, the tiles must be for example wrapped
+  individually in a VRT file before being referenced in the GTI index.
+
+Connection strings
+------------------
+
+The GTI driver accepts different types of connection strings:
+
+* a vector file in GeoPackage format with a ``.gti.gpkg`` extension, or in
+  FlatGeoBuf format with a ``.gti.fgb`` extension, meeting the minimum requirements
+  for a GTI compatible tile index, detailed later.
+
+  For example: ``tileindex.gti.gpkg``
```

Comment:
@rouault I'm 👎 on this. A new kind of format hint is more confusing than helpful.

Reply:
> A new kind of format hint is more confusing than helpful.

Why ? How would the driver be able to quickly identify (that is without opening the dataset) that a GeoPackage file is actually meant to be used with the GTI driver ?

Reply:
Using the "GTI:" suffix might work well for command line usages, but the aim is also that users can for example drag&drop such files in QGIS if needed, and thus it needs to look as a regular file, hence the need for a hint in the filename

Reply:
@sgillies would `GTI:tileindex.gpkg` be more acceptable? I kind of agree that it could be misleading that a file could be opened with an unexpected driver based solely on its name (one could argue that the .gti part of .gti.gpkg is not the extension)

Reply:
@rouault GeoPackage can already be vector or raster. How does QGIS currently determine which driver to use if you drag and drop? Is peeking in the file really not allowed? It's pretty normal behavior for applications like QGIS, yes?

Is "gti" in the filename a signal to QGIS? And then it uses that info to determine whether to flag `GDAL_OF_RASTER` or `GDAL_OF_VECTOR` to `GDALOpenEx()`? GDAL doesn't have a generic opener that dispatches to one or the other yet, does it? I thought you had to know the data model before you call.

Generally, I believe that more and more special cases cause a degradation of user experience. And increase the complication for projects like Rasterio, that's for sure. From my perspective it would be great if we could avoid new special cases.

Reply:
> How does QGIS currently determine which driver to use if you drag and drop?

It largely delegates to GDAL. Here one point I didn't investigate on QGIS side is if he'd prefer the raster driver (GTI) or the vector driver (GPKG), but if using the QGIS browser, it might propose both. So maybe a bit of tuning needed on QGIS side to make it prefer the raster, unless the user explicitly do "open vector layer"

>  Is peeking in the file really not allowed?

generally we avoid that as much as possible for the GDALDriver::Identify() method.  For a SQLite DB, that would require sqlite3_open() (which must be avoided as much as possible because of SQLite locking issues) and some non trivial queries

> Is "gti" in the filename a signal to QGIS? 

no, it is a signal for the OGR GPKG driver *not* to recognize such file as a raster when GDALOpen() in raster mode is attempted on that file.
Cf the following snippet in the Identify() method of the GPKG driver:
```
    if ((poOpenInfo->nOpenFlags & GDAL_OF_RASTER) != 0 &&
        ENDS_WITH_CI(poOpenInfo->pszFilename, ".gti.gpkg"))
    {
        // Handled by GTI driver
        return FALSE;
    }
```

> And increase the complication for projects like Rasterio, that's for sure. 

Rasterio shouldn't have to care about that at all.The end user provides some opaque string "foo.gti.gpkg" to RasterIO open() method which just delegates that to GDALOpen() without having to know more about it. Or am I missing something?

Reply:
I've added an extra commit so that the regular GPKG driver is able to open a GeoPackage tiled raster even if its filename ends with .gti.gpkg, so there not be any functional regression due to the introduction of the GTI driver